common/network-types/src/udp.rs

```diff
@@ -1,152 +1,508 @@
+    ///
+    /// Default is unbounded.
+    pub fn with_queue_size(mut self, queue_size: usize) -> Self {
+        self.queue_size = Some(queue_size);
+        self
     }
 
-    /// Changes the mode of behavior when foreign data (not from the counterparty) is received.
-    /// See [ForeignDataMode] for details.
-    pub fn with_foreign_data_mode(mut self, mode: ForeignDataMode) -> Self {
-        self.foreign_data_mode = mode;
+    /// Sets how many parallel readers/writer sockets should be bound.
+    ///
+    /// Each UDP socket is bound with `SO_REUSEADDR` to facilitate parallel processing
+    /// of read and write operations.
+    ///
+    /// - If some value `n` > 0 is given, the stream will bind `n` sockets.
+    /// - If 0 is given, the number of sockets is determined by [`std::thread::available_parallelism`].
+    /// - If none is given, only a single socket will be created (no parallelism).
+    ///
+    /// Default is none.
+    pub fn with_parallelism(mut self, parallelism: usize) -> Self {
+        self.parallelism = Some(parallelism);
         self
     }
 
-    /// Inner UDP socket.
-    pub fn socket(&self) -> &UdpSocket {
-        &self.sock
+    /// Sets the expected counterparty for data sent/received by the UDP sockets.
+    ///
+    /// If not specified, the counterparty is determined from the first packet received.
+    /// However, no data can be sent up until this point.
+    /// Therefore, the value must be set if data are sent first rather than received.
+    /// If data is expected to be received first, the value does not need to be set.
+    ///
+    /// See [`ConnectedUdpStream`] and [`ForeignDataMode`] for details.
+    ///
+    /// Default is none.
+    pub fn with_counterparty(mut self, counterparty: std::net::SocketAddr) -> Self {
+        self.counterparty = Some(counterparty);
+        self
+    }
+
+    /// Builds the [`ConnectedUdpStream`] with UDP socket(s) bound to `bind_addr`.
+    ///
+    /// The number of sockets bound is determined by [parallelism](UdpStreamBuilder::with_parallelism).
+    /// The returned instance is always ready to receive data.
+    /// It is also ready to send data
+    /// if the [counterparty](UdpStreamBuilder::with_counterparty) has been set.
+    pub fn build<A: std::net::ToSocketAddrs>(self, bind_addr: A) -> std::io::Result<ConnectedUdpStream> {
+        let avail_parallelism = std::thread::available_parallelism()
+            .map(usize::from)
+            .unwrap_or_else(|e| {
+                error!("failed to determine available parallelism, defaulting to 1: {e}");
+                1
+            })
+            / 2; // Each socket gets one RX and one TX task
+        let num_socks = self
+            .parallelism
+            .map(|n| {
+                if n > 0 {
+                    n.max(avail_parallelism)
+                } else {
+                    avail_parallelism
+                }
+            })
+            .unwrap_or(1);
```

Comment:
This reads slightly better:

```suggestion
        let num_socks = self
            .parallelism
            .unwrap_or(1)
            .max(avail_parallelism);
```

Reply:
It is not the same though.
In the case it is `Some(0)` , `num_socks` should have the value of `avail_parallelism`.

Didn't want to bother with a custom enum for parallelism.